src/tbb/market.cpp

```diff
@@ -590,10 +596,13 @@ void market::adjust_demand ( arena& a, int delta, bool mandatory ) {
 void market::process( job& j ) {
     thread_data& td = static_cast<thread_data&>(j);
     // td.my_arena can be dead. Don't access it until arena_in_need is called
-    arena *a = td.my_arena;
+    arena* a = td.my_arena;
     for (int i = 0; i < 2; ++i) {
-        while ( (a = arena_in_need(a)) ) {
-            a->process(td);
+        while ((a = arena_in_need(a, nullptr))) {
+            arena* next_arena{a};
+            do {
+                next_arena = next_arena->process(td);
+            } while (next_arena);
```

Comment:
Why do we need two loops? Does not `arena::process` checks `arena_in_need` and there is no sense to call `arena_in_need` if `arena::process` returned `nullptr`?